### PR TITLE
fix issue that causes illegal diff query

### PR DESCRIPTION
**Pull request by ajtmccarty**
IFC-1013

when aggregating multiple diffs as part of a diff update, if any diffs in the aggregation list overlapped it would cause an illegal diff query to be created in which the `to_time` was before the `from_time`

this fix is to go through the list of diffs to aggregate and remove any that overlap, giving preference to the diff that covers the greater time range

also fixes a small bug in the diff query that could cause a diff for an arbitrary time frame (meaning a `to_time` other than "right now") to not include changes. this is not something a user would encounter unless they are using GraphQL to generate their own diffs with arbitrary time ranges outside of the branch/proposed change process. and I highly doubt anyone is doing that

### CODE REVIEW

File: backend/infrahub/core/diff/coordinator.py [hunk 1]
```diff
@@ -219,6 +218,30 @@ async def recalculate(
...
+            candidate_diff = candidate_diff_pair.diff_branch_diff
+            if previous_diff.to_time <= candidate_diff.from_time:
+                ordered_diffs_no_overlaps.append(candidate_diff_pair)
+                continue
+            previous_interval = previous_diff.time_range
+            candidate_interval = candidate_diff.time_range
+            # keep the diff that covers the larger time frame
+            if candidate_interval > previous_interval:
+                ordered_diffs_no_overlaps[-1] = candidate_diff_pair
+        return ordered_diffs_no_overlaps
```
Comment (ajtmccarty) [inline backend/infrahub/core/diff/coordinator.py:243]: strip out overlapping diffs. when two diffs overlap, keep the one that covers the greater time range

File: backend/infrahub/core/diff/coordinator.py [hunk 2]
```diff
@@ -272,15 +295,17 @@ async def _get_aggregated_enriched_diffs(
         if not partial_enriched_diffs:
             return await self._get_enriched_diff(diff_request=diff_request)
 
-        remaining_diffs = sorted(partial_enriched_diffs, key=lambda d: d.diff_branch_diff.from_time)
+        ordered_diffs = self._get_ordered_diff_pairs(diff_pairs=partial_enriched_diffs, allow_overlap=False)
+        ordered_diff_reprs = [repr(d) for d in ordered_diffs]
+        log.debug(f"Ordered diffs for aggregation: {ordered_diff_reprs}")
```
Comment (ajtmccarty) [inline backend/infrahub/core/diff/coordinator.py:300]: a little logging to help us troubleshoot if this comes up again

File: backend/infrahub/core/diff/coordinator.py [hunk 3]
```diff
@@ -322,26 +347,6 @@ async def _get_enriched_diff(self, diff_request: EnrichedDiffRequest) -> Enriche
...
-        while index < len(sorted_time_ranges) - 1:
-            this_diff = sorted_time_ranges[index]
-            next_diff = sorted_time_ranges[index + 1]
-            if this_diff.to_time < next_diff.from_time:
-                missing_time_ranges.append(TimeRange(from_time=this_diff.to_time, to_time=next_diff.from_time))
-            index += 1
-        if sorted_time_ranges[-1].to_time < to_time:
-            missing_time_ranges.append(TimeRange(from_time=sorted_time_ranges[-1].to_time, to_time=to_time))
-        return missing_time_ranges
-
```
Comment (ajtmccarty) [inline backend/infrahub/core/diff/coordinator.py:344]: dead code

File: backend/infrahub/core/query/diff.py [hunk 4]
```diff
@@ -767,8 +767,10 @@ async def query_init(self, db: InfrahubDatabase, **kwargs: Any) -> None:
         // -------------------------------------
         MATCH diff_rel_path = (root:Root)<-[r_root:IS_PART_OF]-(n:Node)-[r_node]-(p)-[diff_rel {branch: $branch_name}]->(q)
         WHERE (node_field_specifiers_list IS NULL OR [n.uuid, p.name] IN node_field_specifiers_list)
-        AND (from_time <= diff_rel.from < $to_time)
-        AND (diff_rel.to IS NULL OR (from_time <= diff_rel.to < $to_time))
+        AND (
+            (from_time <= diff_rel.from < $to_time)
+            OR (from_time <= diff_rel.to < $to_time)
+        )
```
Comment (ajtmccarty) [inline backend/infrahub/core/query/diff.py:773]: fix in the query to calculate a diff

File: backend/tests/unit/core/diff/test_diff_calculator.py [hunk 5]
```diff
@@ -709,6 +709,12 @@ async def test_relationship_one_property_branch_update(
     assert property_diff.new_value is None
     assert property_diff.action is DiffAction.REMOVED
     assert before_main_change < property_diff.changed_at < after_main_change
+    property_diff = property_diff_by_type[DatabaseEdgeType.IS_PROTECTED]
+    assert property_diff.property_type == DatabaseEdgeType.IS_PROTECTED
+    assert property_diff.previous_value is False
+    assert property_diff.new_value is None
+    assert property_diff.action is DiffAction.REMOVED
+    assert before_main_change < property_diff.changed_at < after_main_change
```
Comment (ajtmccarty) [inline backend/tests/unit/core/diff/test_diff_calculator.py:717]: fix to this unit test after the fix to the diff query

**Review by dgarros — APPROVED** [review]

**Pull request merged**